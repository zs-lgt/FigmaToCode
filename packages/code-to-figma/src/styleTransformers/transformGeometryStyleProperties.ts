import { GeometryProps } from '../types';
import { colorToPaint, colorToRGB } from './transformColors';
import { LayoutStyleProperties } from './transformLayoutStyleProperties';
import { transformSize } from '../helpers/size';

export type ResizeMode = 'contain' | 'cover' | 'stretch' | 'center' | 'repeat' | 'none';

export type GeometryStyleProperties = {
    backgroundColor: string;
    backgroundImage: string | { uri: string } | { default: string };
    backgroundSize: ResizeMode;
    fillStyleId?: string;
};

const backgroundSizeToScaleMode = {
    cover: 'FILL',
    contain: 'FIT',
    repeat: 'TILE',
    center: 'CROP',
    stretch: 'CROP',
    none: 'FILL'
};

export const transformGeometryStyleProperties = (
    property: 'fills' | 'backgrounds',
    style?: Partial<LayoutStyleProperties & GeometryStyleProperties & { strokeStyleId?: string }>,
    imageHash?: string
): GeometryProps => {
    if (!style) {
        return {};
    }

    const fills = [];

    if (style.backgroundColor) {
        fills.push(colorToPaint(style.backgroundColor));
    }

    if (style.backgroundImage) {
        let color;
        try {
            color = colorToPaint(
                typeof style.backgroundImage === 'string'
                    ? style.backgroundImage
                    : 'uri' in style.backgroundImage
                    ? style.backgroundImage.uri
                    : style.backgroundImage.default
            );
        } catch (e) {}
        if (color) {
            fills.push(color);
        } else if (imageHash) {
            if (style.backgroundSize === 'stretch') {
                fills.push({
                    type: 'IMAGE',
                    imageHash,
                    scaleMode: backgroundSizeToScaleMode.stretch,
                    imageTransform: [[transformSize(style.width), 0, 0], [0, transformSize(style.height), 0]]
                });
            } else {
                fills.push({
                    type: 'IMAGE',
                    imageHash,
                    scaleMode: style.backgroundSize
                        ? backgroundSizeToScaleMode[style.backgroundSize]
                        : backgroundSizeToScaleMode.cover
                });
            }
        }
    }

    return {
        ...((fills.length > 0 && { [property]: fills }) || {}),
        ...((style.fillStyleId && { fillStyleId: style.fillStyleId }) || {}),
        ...((style.strokeStyleId && { strokeStyleId: style.strokeStyleId }) || {})
    };
};
