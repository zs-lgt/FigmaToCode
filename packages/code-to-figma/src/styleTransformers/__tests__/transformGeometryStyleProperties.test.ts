import { transformGeometryStyleProperties } from '../transformGeometryStyleProperties';

describe('transformGeometryStyleProperties', () => {
    it('backgroundColor: "red"', () => {
        const result = transformGeometryStyleProperties('fills', { backgroundColor: 'red' });
        expect(result).toMatchSnapshot();
    });

    it('backgroundImage: ./image.png', () => {
        const result = transformGeometryStyleProperties('fills', { backgroundImage: './image.png' });
        expect(result).toMatchSnapshot();
    });

    it('imageHash transforms to fills', () => {
        const result = transformGeometryStyleProperties(
            'fills',
            { backgroundImage: './image.png' },
            'data:image/png;base64,iVBORw0KGgoAAA\n' +
                'ANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4\n' +
                '//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU\n' +
                '5ErkJggg=='
        );
        expect(result).toMatchSnapshot();
    });

    it('imageHash transforms to fills with uri as background image', () => {
        const result = transformGeometryStyleProperties(
            'fills',
            { backgroundImage: { uri: './image.png' } },
            'data:image/png;base64,iVBORw0KGgoAAA\n' +
                'ANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4\n' +
                '//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU\n' +
                '5ErkJggg=='
        );
        expect(result).toMatchSnapshot();
    });

    it('imageHash transforms to fills with default as background image', () => {
        const result = transformGeometryStyleProperties(
            'fills',
            { backgroundImage: { uri: './image.png' } },
            'data:image/png;base64,iVBORw0KGgoAAA\n' +
                'ANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4\n' +
                '//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU\n' +
                '5ErkJggg=='
        );
        expect(result).toMatchSnapshot();
    });

    it('backgroundImage: linear-gradient(#e66465, #9198e5)', () => {
        const result = transformGeometryStyleProperties('fills', {
            backgroundImage: 'linear-gradient(#e66465, #9198e5)'
        });
        expect(result).toMatchSnapshot();
    });
});
