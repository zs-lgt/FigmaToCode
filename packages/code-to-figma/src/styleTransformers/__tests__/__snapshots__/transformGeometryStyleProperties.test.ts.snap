// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformGeometryStyleProperties backgroundColor: "red" 1`] = `
Object {
  "fills": Array [
    Object {
      "color": Object {
        "b": 0,
        "g": 0,
        "r": 1,
      },
      "opacity": 1,
      "type": "SOLID",
    },
  ],
}
`;

exports[`transformGeometryStyleProperties backgroundImage: ./image.png 1`] = `Object {}`;

exports[`transformGeometryStyleProperties backgroundImage: linear-gradient(#e66465, #9198e5) 1`] = `
Object {
  "fills": Array [
    Object {
      "gradientStops": Array [
        Object {
          "color": Object {
            "a": 1,
            "b": 0.396078431372549,
            "g": 0.39215686274509803,
            "r": 0.9019607843137255,
          },
          "position": 0,
        },
        Object {
          "color": Object {
            "a": 1,
            "b": 0.8980392156862745,
            "g": 0.596078431372549,
            "r": 0.5686274509803921,
          },
          "position": 1,
        },
      ],
      "gradientTransform": Array [
        Array [
          6.123233995736766e-17,
          1,
          0,
        ],
        Array [
          -1,
          6.123233995736766e-17,
          1,
        ],
      ],
      "type": "GRADIENT_LINEAR",
    },
  ],
}
`;

exports[`transformGeometryStyleProperties imageHash transforms to fills 1`] = `
Object {
  "fills": Array [
    Object {
      "imageHash": "data:image/png;base64,iVBORw0KGgoAAA
ANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4
//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU
5ErkJggg==",
      "scaleMode": "FILL",
      "type": "IMAGE",
    },
  ],
}
`;

exports[`transformGeometryStyleProperties imageHash transforms to fills with default as background image 1`] = `
Object {
  "fills": Array [
    Object {
      "imageHash": "data:image/png;base64,iVBORw0KGgoAAA
ANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4
//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU
5ErkJggg==",
      "scaleMode": "FILL",
      "type": "IMAGE",
    },
  ],
}
`;

exports[`transformGeometryStyleProperties imageHash transforms to fills with uri as background image 1`] = `
Object {
  "fills": Array [
    Object {
      "imageHash": "data:image/png;base64,iVBORw0KGgoAAA
ANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4
//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU
5ErkJggg==",
      "scaleMode": "FILL",
      "type": "IMAGE",
    },
  ],
}
`;
